src/assignments/phat/lesson_5_answers.clj

```diff
@@ -0,0 +1,182 @@
+
+(def file "src/assignments/phat/tmp/somefile.txt")
+
+(def regex-patterns {:url "^https?://[^/]+/([^/]+)/.*$"
+                     :words "\\w+"
+                     :digit "\\d+"})
+
+(defn not-lazy-grep
+  "Returns a sequence of matches of pattern in a file"
+  [filepath regex]
+  (assert (.exists (io/file filepath))
+          "File path does not exist")
+  (let [re (re-pattern regex)]
+    (with-open [rdr (io/reader filepath)]
+      (re-seq re (slurp rdr)))))
+
+(not-lazy-grep file (:digit regex-patterns))
+
+(defn lazy-grep 
+  "Returns a lazy sequence of matches of a regex pattern from a file"
+  [filepath regex]
+  (assert (.exists (io/file filepath))
+          "File path doesn't exist")
+  (let [re (re-pattern regex)]
+    ;; return a sequence of matches for each line in the file
+    (with-open [rdr (io/reader filepath)] 
+      (->> (line-seq rdr)
+           (map #(re-seq re %))
+           (remove nil?)
+           (flatten)
+           (doall)))))
+
+
+;; i'm not too sure with the solution here since the doall forces the sequence resulted from map to be relized
+;; but the function does return a lazy sequence!!
+(class (lazy-grep file "\\w+"))
+
+
+;; the latter is more useful when we need to process a very large file as the function does not need to realize the input at the beginning
+
+
+;; for extra difficulty, make it return a lazy sequence of matches instead. What difficulties does this present? Which version is better/more useful? 
+
+
+;; Can you implement the former in terms of the latter? Discuss.
+;; what if we wrap the former function in a lazy sequence? does this answer the question?
+
+(defn lz-not-lazy-grep
+  "Returns a lz-sequence of matches of pattern in a file"
+  [filepath regex]
+  (assert (.exists (io/file filepath))
+          "File path does not exist")
+  (let [re (re-pattern regex)]
+    (lazy-seq
+     (with-open [rdr (io/reader filepath)]
+       (re-seq re (slurp rdr))))))
+
+(class (lz-not-lazy-grep file "\\w+"))
+
+;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
+;; Day 1 - Advent of Code 2019
+;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
+
+;;;;;;;;;
+;; part 1
+;; problem description: https://adventofcode.com/2019/day/1
+;; can not download input directly from this url: https://adventofcode.com/2019/day/1/input
+;; so I have to save the input in a text file
+(def inputfile "src/assignments/phat/tmp/adventofcode-day1-input.txt")
+
+(defn calculate-fuel
+  [mass]
+  (- (int (/ mass 3)) 2))
+
+(defn calculate-fuel-sum
+  "Takes an input file including a sequence of module's masses
+   and returns sum of fuel requirements for all of the modules"
+  [input]
+  (let [module-masses (map read-string (lazy-grep input "\\d+"))]
+    (reduce + (map calculate-fuel module-masses))))
+
+(calculate-fuel-sum inputfile)
+
+;;;;;;;;;
+;; part 2
+
+(defn total-module-fuel
+  "Takes a module's mass and returns total fuel requirement for that module"
+  [mass]
+  (loop [fuel (calculate-fuel mass)
```

Comment:
You mean the reduced function is the calculate-fuel? I have spent all day trying to rewrite total-module-fuel using reduce but haven't figured out how to solve it. 

My first thought is to generate a sequence of reduced values from calculate-fuel then apply "reduce +" to it but I'm facing a problem generating a sequence in which an element is the output of a function taking the previous element as an input. I'm not sure if I'm on the right track here.

 Can you give me a hint? :)

My first thought to this problem is to recursively apply a function to reduce a value until it reaches zero. Other than this, I don't know

Reply:
I think I misread your code a little bit,
so the "nice" solution is probably not what
I originally thought; but I'm trying to encourage
you into trying to think in terms of filters,
sequences etc on immutable values, rather than
imperative looping constructs.  Consider this code:

You currently have this:
```
(total-module-fuel 1000)
;; => 483
```

Look and try to understand what this code does:
```
(reduce +
        (take-while pos?
                    (iterate calculate-fuel 1000)))
;; => 1483
```
or, written this way:
```
(->> (iterate calculate-fuel 1000)
     (take-while pos?)
     (reduce +))
```

What do you think of that?

Reply:
ohhh cool, iterate is exactly what I was looking for. I got it. Thank you very much!